0003-interledger-protocol/0003-interledger-protocol.md

```diff
@@ -243,10 +254,105 @@ Not all ledgers support held transfers. In the case of a ledger that doesn't, th
+
+Unless otherwise specified, `data` SHOULD be encoded as UTF-8.
+
+### ILP Error Codes
+
+Inspired by [HTTP Status Codes](https://tools.ietf.org/html/rfc2616#section-10), ILP errors are categorized based on the intended behavior of the caller when they get the given error.
+
+#### F__ - Final Error
+
+Final errors indicate that the payment is invalid and should not be retried unless the details are changed.
+
+| Code | Name | Description |
+|---|---|---|
+| **F00** | **Bad Request** | Generic sender error. |
+| **F01** | **Invalid Packet** | The ILP packet was syntactically invalid. |
+| **F02** | **Unreachable** | There was no way to forward the payment, because the destination ILP address was wrong or the connector does not have a route to the destination. |
+| **F03** | **Invalid Amount** | The amount is invalid, for example it contains more digits of precision than are available on the destination ledger or the amount is greater than the total amount of the given asset in existence. |
+| **F04** | **Insufficient Destination Amount** | The receiver deemed the amount insufficient, for example you tried to pay a $100 invoice with $10. |
+| **F05** | **Wrong Condition** | The receiver generated a different condition and cannot fulfill the payment. |
+| **F06** | **Unexpected Payment** | The receiver was not expecting a payment like this (the memo and destination address don't make sense in that combination, for example if the receiver does not understand the transport protocol used) |
+| **F07** | **Cannot Receive** | The receiver is unable to accept this payment due to a constraint. For example, the payment would put the receiver above its maximum account balance. |
+| **F99** | **Application Error** | Reserved for application layer protocols. Applications MAY use names other than `Application Error`. |
+
+#### T__ - Temporary Error
+
+Temporary errors indicate a failure on the part of the receiver or an intermediary system that is unexpected or likely to be resolved soon. Senders SHOULD retry the same payment again, possibly after a short delay.
+
+| Code | Name | Description |
+|---|---|---|
+| **T00** | **Internal Error** | A generic unexpected exception. This usually indicates a bug or unhandled error case. |
+| **T01** | **Ledger Unreachable** | The connector has a route or partial route to the destination but was unable to reach the next ledger. Try again later. |
+| **T02** | **Ledger Busy** | The ledger is rejecting requests due to overloading. Try again later. |
+| **T03** | **Connector Busy** | The connector is rejecting requests due to overloading. Try again later. |
+| **T04** | **Insufficient Liquidity** | The connector would like to fulfill your request, but it doesn't currently have enough money. Try again later. |
+| **T05** | **Rate Limited** | The sender is sending too many payments and is being rate-limited by a ledger or connector. |
```

Comment:
Not just "the sender", but other entities (for example, a connector in the middle of the path) might be the one being rate limited. It might also be a limit on payments received, not sent, right?